### ISSUE TITLE
Abseil linking errors when building simple C++ project using code from main

### ISSUE THREAD

**Issue by palade** [protocolbuffers/protobuf#11920]
**What version of protobuf and what language are you using?**
Version: main
Language: C++

**What operating system (Linux, Windows, ...) and version?**
Linux

**What runtime / compiler are you using (e.g., python version or gcc version)**

g++ (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 

**What did you do?**

1. Build the protobuf library using the main branch
git clone https://github.com/protocolbuffers/protobuf.git
cd protobuf
git submodule update --init --recursive
cd cmake
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/home/andrei/.local ..
cmake --build . --target install

2. Create a simple C++ protobuf project:

protobuffer.cpp
```
#include <iostream>

#include "../build/person.pb.h"

int main(int argc, char** argv) {
 
    tutorial::Person person;

    person.set_name("JOHN DONE");

    return 0;
}
```

person.proto
```
syntax = "proto3";

package tutorial;

message Person {
    string name = 1;
}

```

Makefile
```
LDFLAGS = /home/andrei/.local/lib/libprotobuf.a
INCLUDES = /home/andrei/.local/include

PROTOS_PATH = protos
BUILD_PATH = build
SOURCE_PATH = src

all:
    make clean
    mkdir $(BUILD_PATH)
    protoc -I=$(PROTOS_PATH) --cpp_out=$(BUILD_PATH) person.proto
    g++ -I$(INCLUDES) -g -Wall $(SOURCE_PATH)/protobuffer.cpp $(BUILD_PATH)/person.pb.cc -o $(BUILD_PATH)/protobuffer $(LDFLAGS)

clean:
    rm -rf $(BUILD_PATH)

```

**What did you expect to see**

That the project successfully builds.

**What did you see instead?**

Loads of errors like below:
```
/usr/bin/ld: extension_set.cc:(.text+0x100fd): undefined reference to `absl::lts_20230125::log_internal::LogMessageFatal::~LogMessageFatal()'
/usr/bin/ld: extension_set.cc:(.text+0x10115): undefined reference to `absl::lts_20230125::log_internal::LogMessageFatal::~LogMessageFatal()'
```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment**

These errors do not show up when I'm using the source code from the latest release.

**Comment by jonaski** [comment]
What was the solution here? I'm struggling with the same linking error both with Visual Studio 2022 and MinGW-w64.

**Comment by Yl4682B4** [comment]
> What was the solution here? I'm struggling with the same linking error both with Visual Studio 2022 and MinGW-w64.

I met the problem too, and fixed it by switching branch from master to v21.12, which has no dependency on absl

**Comment by jonaski** [comment]
I worked around the linking issues on MinGW and MSVC by explicitly linking with abseil libraries when compiling my program.

```
if(WIN32 AND Protobuf_VERSION VERSION_GREATER_EQUAL 4.22.0)
  if (MSVC)
    target_link_libraries(libstrawberry-tagreader PRIVATE abseil_dll)
  else()
    target_link_libraries(libstrawberry-tagreader PRIVATE absl_log_internal_message absl_log_internal_check_op)
  endif()
endif()
```
